### PR TITLE
Update repository with the Qibo code for Hash preimages.

### PR DESCRIPTION
**Pull request by igres26**
TODO: Complete README.md files

### CODE REVIEW

File: circuit-specifications/blake_full_main.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    print('QASM file containing all gates created\n')
+    print('The circuit specifications are:\n')
+    print('-'*30)
+    print(circuit.summary)
+    print('-'*30)
+
+
+if __name__ == "__main__":
+    main()
+
```
Comment (scarrazza) [inline circuit-specifications/blake_full_main.py]: There is an empty space in the new line.

File: circuit-specifications/blake_toy_main.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    print('QASM file containing all gates created.\n')
+    print('The circuit specifications are:\n')
+    print('-'*30)
+    print(circuit.summary)
+    print('-'*30)
+
+
+if __name__ == "__main__":
+    main()
+
```
Comment (scarrazza) [inline circuit-specifications/blake_toy_main.py]: Also here.

File: circuit-specifications/sponge_full_main.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    print('QASM file containing all gates created.\n')
+    print('The circuit specifications are:\n')
+    print('-'*30)
+    print(circuit.summary)
+    print('-'*30)
+
+
+if __name__ == "__main__":
+    main()
+
```
Comment (scarrazza) [inline circuit-specifications/sponge_full_main.py]: Also here.

File: circuit-specifications/sponge_toy_main.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    print('QASM file containing all gates created.\n')
+    print('The circuit specifications are:\n')
+    print('-'*30)
+    print(circuit.summary)
+    print('-'*30)
+
+
+if __name__ == "__main__":
+    main()
+
```
Comment (scarrazza) [inline circuit-specifications/sponge_toy_main.py]: Also here.

File: functions.py [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+        result (dict): counts of the output generated by the algorithm
+    """
+    A, B, C, D, x, ancilla, circuit, qubits = create_qc(q)
+    c1, c2, c3, c4 = initial_step(q, constant_1, constant_2, rot)
+    c = []
+    c.append(c1)
+    c.append(c2)
+    c.append(c3)
+    c.append(c4)
+    circuit.add(start_grover(A+B, ancilla))
```
Comment (scarrazza) [inline functions.py]: I think the error is missing:
```suggestion
    circuit.add(start_grover(A+B, ancilla, err))
```

**Comment by scarrazza** [comment]
I have tested the default `main.py` in a clean conda python 3.8 environment and it works.
When you create a new environment it seems to me that the only requirement is `pip install qibo` before running the code, so you can probably simply add this line in the readme.

**Comment by igres26** [comment]
Both `README.md` files are updated with the information on how to run the code.

File: README.md [hunk 6]
```diff
@@ -1,19 +1,29 @@
 [![DOI](https://zenodo.org/badge/265590690.svg)](https://zenodo.org/badge/latestdoi/265590690)
 # Quantum Search for Scaled Hash Function Preimages
-### OpenQASM 2.0 files
 
-This repository contains examples for the Grover search implementations of scaled down versions of Chacha based Sponge hash functions and Blake2 construction.
+This repository contains code to reproduce the result presented in the paper: Quantum Search for Scaled Hash Function Preimages.
 
-The OpenQASM convertion has been possible thanks to the Qibo quantum simulation language.
+The scripts presented use [Qibo](https://github.com/Quantum-TII/qibo), a Python library for classical simulation of quantum circuits. In order to properly run the scripts, please start by installing Qibo to your Python enviroment with `pip install qibo`.
```
Comment (scarrazza) [inline README.md]: I would suggest to move the pip installation to a new line, e.g.:
```
pip install qibo
```

**Comment by scarrazza** [comment]
In general this looks good, but in terms of maintainability I would suggest to include a CI job which tests the code, let me try to add some template code so if you agree you can add some minimal tests, to make sure the code works.

**Comment by igres26** [comment]
Okay! I think the plain and collisions test are enough for `main.py` as noise is too costly to calculate.

Adding tests to the circuit-specifications folder would mean generating the qasm files?

**Comment by scarrazza** [comment]
> Adding tests to the circuit-specifications folder would mean generating the qasm files?

not really, just edit the `.github/workflow/test.yaml` and include the python execution functions you would like to try, if you want it is also possible to test for the right result.

**Comment by igres26** [comment]
I have added tests for everything now. If everything works it might be done.

**Comment by scarrazza** [comment]
Ok, great, then lets merge and create a new release.

**Comment by igres26** [comment]
Great. I will merge and create a new release.

**Pull request merged**